overlord/snapstate/snapstate.go

```diff
@@ -710,6 +724,26 @@ var generateSnapdWrappers = backend.GenerateSnapdWrappers
 // For snapd snap updates this will also rerun wrappers generation to fully
 // catch up with any change.
 func FinishRestart(task *state.Task, snapsup *SnapSetup) (err error) {
+	// For classic we have not forced a reboot, so we need to look at the
+	// boot id to check if the reboot has already happened or not. If not,
+	// we return with a Hold error.
+	if release.OnClassic {
+		// boot-id will be present only if a reboot was required,
+		// otherwise we continue down the function.
+		var changeBootId string
+		if err := task.Change().Get("boot-id", &changeBootId); err == nil {
+			currentBootID, err := osutil.BootID()
+			if err != nil {
+				return err
+			}
+			if currentBootID == changeBootId {
+				task.Logf("Waiting for manual restart...")
+				return &state.Hold{Reason: "waiting for user to reboot"}
+			}
+			logger.Debugf("restart already happened for change %s",
+				task.Change().ID())
+		}
+	}
```

Comment:
this should become also some form of overlord/restart helper

Reply:
Done now